### ISSUE TITLE
Wrong results when using tunable number of layers

### ISSUE THREAD

**Issue by Ali-Ai** [keras-team/keras-tuner#272]
The num_layer parameter is changing but always the maximum number of layers are produced. I have brought the example below:

The relevant part of my code is:
----------------
def build_model(hp):
    model = Sequential()
    model.add(Dense(units=hp.Int('hl1-units', min_value=16, max_value=512, step=16),
                    activity_regularizer=regularizers.l1(hp.Float('regl1',
                                                                  min_value=1e-4,
                                                                  max_value=1e-1,
                                                                  sampling='LOG',
                                                                  default=1e-3)),
                    input_shape=(input_dim, )))
    model.add(Activation(hp.Choice('hl1-activation',
                                   values=['sigmoid', 'relu', 'tanh'], default='relu')))

    for i in range(hp.Int('num_layers', 1, 4)):
        model.add(Dense(units=hp.Int('hl' + str(i + 2) + '-units',
                                     min_value=16,
                                     max_value=512,
                                     step=16)))
        model.add(BatchNormalization())
        model.add(Activation(hp.Choice('hl' + str(i + 2) + '-activation',
                                       values=['sigmoid', 'relu', 'tanh'], default='relu')))
        model.add(Dropout(rate=hp.Float('hl' + str(i + 2) + '-dropout',
                                        min_value=0.0,
                                        max_value=0.5,
                                        step=0.1)))

    model.add(Dense(output_dim, activation='sigmoid'))

    model.compile(optimizer=Adam(lr=1e-2,
                                 decay=hp.Float('decay',
                                                min_value=1e-5,
                                                max_value=1e-2,
                                                sampling='LOG',
                                                default=1e-3)),
                  loss='binary_crossentropy', metrics=[utils.exact_accuracy])
    return model

ht_dir = 'hypetune'
path = os.path.join(ht_dir, f"{model_name}")
shutil.rmtree(path, ignore_errors=True)

objective = kerastuner.Objective("val_loss", direction="min")
tuner = RandomSearch(
                    build_model,
                    objective=objective,
                    max_trials=N_TRIALS,
                    executions_per_trial=1,
                    directory=ht_dir,
                    project_name=model_name)

tuner.search_space_summary()
hist = tuner.search(train_dataset,
                    train_labels,
                    verbose=2,
                    epochs=EPOCHS,
                    batch_size=BS,
                    callbacks=[early_stop_callback],
                    validation_data=(test_dataset, test_labels))
-------------
But during running I got following output:
 
Train on 159200 samples, validate on 39800 samples
Epoch 1/3
159200/159200 - 9s - loss: 0.4995 - exact_accuracy: 0.5300 - val_loss: 0.4821 - val_exact_accuracy: 0.5373
Epoch 2/3
159200/159200 - 8s - loss: 0.4659 - exact_accuracy: 0.5618 - val_loss: 0.4601 - val_exact_accuracy: 0.5891
Epoch 3/3
159200/159200 - 8s - loss: 0.4668 - exact_accuracy: 0.5636 - val_loss: 0.4675 - val_exact_accuracy: 0.5469
[Trial complete]
[Trial summary]
 |-Trial ID: e6c0939a8f68cb99c94ded37273cc670
 |-Score: 0.46014539468228516
 |-Best step: 0
 > Hyperparameters:
 |-decay: 1.0094176563068276e-05
 |-hl1-activation: sigmoid
 |-hl1-units: 400
 |-hl2-activation: sigmoid
 |-hl2-dropout: 0.1
 |-hl2-units: 16
 |-hl3-activation: tanh
 |-hl3-dropout: 0.30000000000000004
 |-hl3-units: 240
 |-hl4-activation: tanh
 |-hl4-dropout: 0.5
 |-hl4-units: 80
 |-hl5-activation: relu
 |-hl5-dropout: 0.0
 |-hl5-units: 96
 |-num_layers: 1
 |-regl1: 0.0065076212574787734

in which the trial summary says that the network is always producing 5 hidden layers whereas the correct number of hidden layers due to the num_layers parameter in the summary should be two hidden layers.

**Comment by ben-arnao** [comment]
Parameters are always optimized on even when they are not actually used. Ie. in random search, layer 5 dropout will still be assigned a random value even when a num_layers variable is only 3.

The build_model function never uses the layer 5 dropout variable, but it is still created, the program just isn't able to know that it is not used at the time the configuration is created.

There's been some discussion on how to solve this. I think there was a merge recently that made a conditional support better. Maybe you can look into this and see if it helps your problem.

However in my fork i deal with things a bit differently by more or less building the model prior to finalizing the configuration and keeping tracking of which parameters are actually pulled from (ie. correctly only showing the active params in your post above).

See #242
